src/backend/metal/src/window.rs

```diff
@@ -355,13 +375,97 @@ impl hal::Surface<Backend> for Surface {
             device_caps.os_is_mac && device_caps.has_version_at_least(10, 13);
 
         let present_modes = if can_set_display_sync {
-            vec![hal::PresentMode::Fifo, hal::PresentMode::Immediate]
+            vec![PresentMode::Fifo, PresentMode::Immediate]
         } else {
-            vec![hal::PresentMode::Fifo]
+            vec![PresentMode::Fifo]
         };
 
         (caps, Some(formats), present_modes)
     }
+
+        /// Set up the swapchain associated with the surface to have the given format.
+    unsafe fn configure_swapchain(
+        &mut self, device: &Device, config: SurfaceSwapchainConfig
+    ) -> Result<(), CreationError> {
+        info!("configure_swapchain {:?}", config);
+
+        let caps = &device.shared.private_caps;
+        let mtl_format = caps
+            .map_format(config.format)
+            .expect("unsupported backbuffer format");
+        self.swapchain_format = mtl_format;
+
+        let render_layer_borrow = self.inner.render_layer.lock();
+        let render_layer = *render_layer_borrow;
+        let framebuffer_only = true;
+        let display_sync = config.present_mode != PresentMode::Immediate;
+        let is_mac = caps.os_is_mac;
+        let can_set_next_drawable_timeout = if is_mac {
+            caps.has_version_at_least(10, 13)
+        } else {
+            caps.has_version_at_least(11, 0)
+        };
+        let can_set_display_sync = is_mac && caps.has_version_at_least(10, 13);
+        let drawable_size = {
+            let bounds: CGRect = msg_send![render_layer, bounds];
```

Comment:
Can this be changed to `self.inner.dimensions()` to take scaling into account?

Reply:
I tried that and got incorrect scaling on the quad example, didn't investigate further

Reply:
This doesn't look right to me. `bounds` returns the drawable size in logical pixels. `drawableSize` ought to be used to get the drawable size in physical pixels. Even if we change this, it appears that this is trying to set the drawable size to the same value that it already is. I'm not sure what was intended here.

Shouldn't we be storing the new extents of the swapchain image in `SurfaceSwapchainConfig` and be setting the drawable size to that instead? Currently, there doesn't seem to be any way to specify the extents of the swapchain.

Reply:
Quad is using logical pixels for everything. Which is why setting the drawable size to `bounds` (logical size) works for quad.